docs/advanced.dj

```diff
@@ -0,0 +1,505 @@
+- `[1, 2, 3] | tobytes as [$x,  $y] | $x, $y --> 1 2`
+
+When using a filter `(f)` as object key in a pattern, then
+`f` is run with the input that was matched
+by its parent object pattern, not
+by the whole pattern.
+For example,
+`[{"k": "a", "a": 1}] as [{(.k): $x}] | $x --> 1`
+
+This is equivalent to:
+
+```
+[{"k": "a", "a": 1}] |
+.[0]          as $p0 |
+$p0[$p0 | .k] as $x  |
+$x
+--> 1
+```
+
+Here, we can see that `(.k)` is run with the input `$p0`, which is
+the value that the parent object pattern of `(.k)`, namely `{(.k): $x}`,
+is trying to match.
+Compare this with the following _wrong_ transformation, where
+`(.k)` would be run with the input matched by the whole pattern:
+
+```
+[{"k": "a", "a": 1}] |
+try (
+.[0]     as $p0 |
+$p0[.k]  as $x  | # fails here because .k is run with whole input
+$x
+) catch "fail" --> "fail"
+```
+:::
+
+
+## Modules
+
+jq allows dividing programs into multiple files that are called *modules*.
+
+At the beginning of any jq module, there is a module header that consists of
+a (potentially empty) sequence of instructions listed in this section.
+The module header is then followed by a sequence of [definitions](#def).
+Finally, the main module that is called from the command-line interface
+(via [`--from-file`](#--from-file) or inline)
+must contain a single filter at the end, which is the filter that is executed.
+
+All `include`/`import` instructions search for files as explained in the
+[search paths](#search-paths) section.
+
+### Module metadata
+
+The instruction `module meta;` sets the metadata of the current module to
+the output of `meta`, where `meta` is a filter.
+This instruction may occur
+only at the beginning of the module header and
+only once.
+
+For example, `module "My module"; 1 --> 1`.
+
+::: Compatibility
+jaq ignores this instruction, whereas
+`jq` uses it to provides the output of `meta` via the `modulemeta` filter.
+:::
+
+### Module inclusion
+
+The instructions
+`include "mod";` and
+`include "mod" meta;`
+make all filters defined in the module `mod.jq` accessible in the current module.
+
+For example, if `foo.jq` in the current working directory contains `def bar: 1;`, then
+`jaq -L . -n 'include "foo"; bar'` yields `1`.
+
+### Module import
+
+The instructions
+`import "mod" as name;` and
+`import "mod" as name meta;`
+make all definitions in the module `mod.jq` accessible in the current module
+with the prefix `name::`.
+
+For example, if `foo.jq` in the current working directory contains `def bar: 1;`, then
+`jaq -L . -n 'import "foo" as myfoo; myfoo::bar'` yields `1`.
+
+### Data import
+
+The instructions
+`import "data" as name;` and
+`import "data" as name meta;`
+load all JSON values in `data.json` to an array,
+bind it to the variable `$name`, and
+make it accessible in the current module.
+
+For example, if `foo.json` in the current working directory contains `1 2 3`, then
+`jaq -L . -n 'import "foo.json" as $myfoo; $myfoo'` yields `1 2 3`.
```

Comment:
yields `[1, 2, 3]`?

Reply:
also noticed that jq/gojq don't like `.json` for data import but for jaq support either way both for module and data import, intentional i guess?